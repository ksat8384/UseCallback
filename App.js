import React, {useState, useCallback} from 'react';
import {SafeAreaView, StyleSheet, Text, TextInput, Button} from 'react-native';
import List from './src/components/List';
import uuid from 'react-native-uuid';

const USERS = [
  {id: 'a', name: 'Batman'},
  {id: 'b', name: 'Superman'},
  {id: 'c', name: 'Ironman'},
  {id: 'i', name: 'Spiderman'},
];

const App = () => {
  console.log('Render : App');
  const [text, setText] = useState('');
  const [users, setUsers] = useState(USERS);

  const handleText = text => {
    setText(text);
  };

  const handleAddUser = () => {
    setUsers(users.concat({id: uuid.v1(), name: text}));
  };

  /**
   *
   * Whenever the App component re-renders after user types into input field, the
   * handleRemove handler function in the App gets re-defined. A new callback handler
   * will be passed to List component which leads to re-render of List and ListItem.
   *
   * If the user state changes by adding or removing an item from the list, the handler function gets
   * re-defined and the child components should re-render. However, if someone only types
   * into the input field, the function doesn't get re-defined and stays intact. Therefore,
   * the child components don't receive changed props and won't re-render for this case
   *
   * Memoized callback function to avoid re-defining this function everytime App re-renders
   *
   * This function only gets re-defined if any of its dependencies in the dependency array changes
   *
   */
  const handleRemove = useCallback(
    id => {
      setUsers(users.filter(user => user.id !== id));
    },
    [users],
  );

  return (
    <SafeAreaView style={styles.container}>
      <TextInput
        style={styles.input}
        placeholder="Enter a new user to add here..."
        value={text}
        onChangeText={handleText}
      />

      <Button title="Add User" onPress={handleAddUser} />

      <List list={users} onRemove={handleRemove} />
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 20,
  },
  input: {
    backgroundColor: 'white',
    borderBottomColor: '#000000',
    borderBottomWidth: 1,
    marginBottom: 10,
  },
});

export default App;
