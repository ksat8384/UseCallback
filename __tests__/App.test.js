import React from "react";
import App from "../App";
import {
  render,
  cleanup,
  fireEvent,
  waitFor,
  within,
} from "@testing-library/react-native";

describe("App Component", () => {
  let component;
  beforeEach(() => {
    component = render(<App />);
  });
  afterEach(cleanup);

  it("renders correctly", () => {
    const { toJSON } = component;
    expect(toJSON()).toMatchSnapshot();
  });

  it("renders input field, button and list correctly", () => {
    const { findByTestId, findByText } = component;

    const input = findByTestId("input");
    const addUserButton = findByText("Add User");
    const list = findByTestId("flat-list");

    expect(input).toBeTruthy();
    expect(addUserButton).toBeTruthy();
    expect(list).toBeTruthy();
  });

  it("Adds the user to the list", async () => {
    const { getByTestId } = component;

    const mockInputText = "Hulk";

    const input = getByTestId("input");
    const addUserButton = getByTestId("add-user");

    fireEvent.changeText(input, mockInputText);
    fireEvent.press(addUserButton);

    await waitFor(() => {
      const list = within(getByTestId("flat-list"));
      expect(list.getByText("Hulk")).toBeTruthy();
    });
  });
});
