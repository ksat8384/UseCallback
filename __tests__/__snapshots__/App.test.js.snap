// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
      "padding": 20,
    }
  }
>
  <TextInput
    onChangeText={[Function]}
    placeholder="Enter a new user to add here..."
    style={
      Object {
        "backgroundColor": "white",
        "borderBottomColor": "#000000",
        "borderBottomWidth": 1,
        "marginBottom": 10,
      }
    }
    testID="input"
    value=""
  />
  <View
    accessibilityRole="button"
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="add-user"
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Add User
      </Text>
    </View>
  </View>
  <RCTScrollView
    data={
      Array [
        Object {
          "id": "a",
          "name": "Batman",
        },
        Object {
          "id": "b",
          "name": "Superman",
        },
        Object {
          "id": "c",
          "name": "Ironman",
        },
        Object {
          "id": "i",
          "name": "Spiderman",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    testID="flat-list"
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#00ffff",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginVertical": 8,
              "padding": 20,
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 32,
              }
            }
          >
            Batman
          </Text>
          <View
            accessibilityRole="button"
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="remove"
          >
            <View
              style={
                Array [
                  Object {},
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#007AFF",
                      "fontSize": 18,
                      "margin": 8,
                      "textAlign": "center",
                    },
                  ]
                }
              >
                Remove
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#00ffff",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginVertical": 8,
              "padding": 20,
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 32,
              }
            }
          >
            Superman
          </Text>
          <View
            accessibilityRole="button"
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="remove"
          >
            <View
              style={
                Array [
                  Object {},
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#007AFF",
                      "fontSize": 18,
                      "margin": 8,
                      "textAlign": "center",
                    },
                  ]
                }
              >
                Remove
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#00ffff",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginVertical": 8,
              "padding": 20,
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 32,
              }
            }
          >
            Ironman
          </Text>
          <View
            accessibilityRole="button"
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="remove"
          >
            <View
              style={
                Array [
                  Object {},
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#007AFF",
                      "fontSize": 18,
                      "margin": 8,
                      "textAlign": "center",
                    },
                  ]
                }
              >
                Remove
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#00ffff",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginVertical": 8,
              "padding": 20,
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 32,
              }
            }
          >
            Spiderman
          </Text>
          <View
            accessibilityRole="button"
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="remove"
          >
            <View
              style={
                Array [
                  Object {},
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#007AFF",
                      "fontSize": 18,
                      "margin": 8,
                      "textAlign": "center",
                    },
                  ]
                }
              >
                Remove
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
